plugin/plugins/telemetry/telemetry.go

```diff
@@ -0,0 +1,538 @@
+}
+
+var Plugins = []plugin.Plugin{
+	&telemetryPlugin{},
+}
+
+type telemetryPlugin struct {
+	uuidFilename string
+	mode         pluginMode
+	endpoint     string
+	runOnce      bool
+	sendDelay    time.Duration
+
+	node      *core.IpfsNode
+	config    *config.Config
+	event     *LogEvent
+	startTime time.Time
+}
+
+func (p *telemetryPlugin) Name() string {
+	return "telemetry"
+}
+
+func (p *telemetryPlugin) Version() string {
+	return "0.0.1"
+}
+
+func readFromConfig(cfg interface{}, key string) string {
+	if cfg == nil {
+		return ""
+	}
+
+	pcfg, ok := cfg.(map[string]interface{})
+	if !ok {
+		return ""
+	}
+
+	val, ok := pcfg[key].(string)
+	if !ok {
+		return ""
+	}
+	return val
+}
+
+func (p *telemetryPlugin) Init(env *plugin.Environment) error {
+	// logging.SetLogLevel("telemetry", "DEBUG")
+	log.Debug("telemetry plugin Init()")
+	p.event = &LogEvent{}
+	p.startTime = time.Now()
+
+	repoPath := env.Repo
+	p.uuidFilename = path.Join(repoPath, uuidFilename)
+
+	v := os.Getenv(modeEnvVar)
+	if v != "" {
+		log.Debug("mode set from env-var")
+	} else if pmode := readFromConfig(env.Config, "Mode"); pmode != "" {
+		v = pmode
+		log.Debug("mode set from config")
+	}
+
+	// read "Delay" from the config. Parse as duration. Set p.delay to it
+	// or set default.
+	if delayStr := readFromConfig(env.Config, "Delay"); delayStr != "" {
+		delay, err := time.ParseDuration(delayStr)
+		if err != nil {
+			log.Debug("sendDelay set from default")
+			p.sendDelay = sendDelay
+		} else {
+			log.Debug("sendDelay set from config")
+			p.sendDelay = delay
+		}
+	} else {
+		log.Debug("sendDelay set from default")
+		p.sendDelay = sendDelay
+	}
+
+	p.endpoint = endpoint
+	if ep := readFromConfig(env.Config, "Endpoint"); ep != "" {
+		log.Debug("endpoint set from config", ep)
+		p.endpoint = ep
+	}
+
+	switch v {
+	case "optout":
+		p.mode = modeOptOut
+		return nil
+	case "info":
+		p.mode = modeInfo
+	default:
+		p.mode = modeOptIn
+	}
+	log.Debug("telemetry mode: ", p.mode)
+	return nil
+}
+
+func (p *telemetryPlugin) loadUUID() error {
+	// Generate or read our UUID from disk
+	b, err := os.ReadFile(p.uuidFilename)
+	if err == nil {
+		v := string(b)
+		v = strings.TrimSpace(v)
+		uid, err := uuid.Parse(v)
+		if err != nil {
+			log.Errorf("cannot parse telemetry uuid: %s", err)
+			return err
+		}
+		log.Debugf("uuid read from disk %s", uid)
+		p.event.UUID = uid.String()
+		return nil
+	} else if os.IsNotExist(err) {
+		uid, err := uuid.NewRandom()
+		if err != nil {
+			log.Errorf("cannot generate telemetry uuid: %s", err)
+			return err
+		}
+		p.event.UUID = uid.String()
+		p.mode = modeInfo
+		log.Debugf("new telemetry UUID %s. Mode set to Info", uid)
+
+		// Write the UUID to disk
+		if err := os.WriteFile(p.uuidFilename, []byte(p.event.UUID), 0600); err != nil {
+			log.Errorf("cannot write telemetry uuid: %s", err)
+			return err
+		}
+	} else {
+		log.Errorf("error reading telemetry uuid from disk: %s", err)
+		return err
+	}
+	return nil
```

Comment:
Minor style suggestion (https://go.dev/wiki/CodeReviewComments#indent-error-flow)


```suggestion
	if err != nil {
		if !os.IsNotExist(err) {
			log.Errorf("error reading telemetry uuid from disk: %s", err)
			return err
		}
		uid, err := uuid.NewRandom()
		if err != nil {
			log.Errorf("cannot generate telemetry uuid: %s", err)
			return err
		}
		p.event.UUID = uid.String()
		p.mode = modeInfo
		log.Debugf("new telemetry UUID %s. Mode set to Info", uid)

		// Write the UUID to disk
		if err := os.WriteFile(p.uuidFilename, []byte(p.event.UUID), 0600); err != nil {
			log.Errorf("cannot write telemetry uuid: %s", err)
			return err
		}
		return nil
	}

	v := string(b)
	v = strings.TrimSpace(v)
	uid, err := uuid.Parse(v)
	if err != nil {
		log.Errorf("cannot parse telemetry uuid: %s", err)
		return err
	}
	log.Debugf("uuid read from disk %s", uid)
	p.event.UUID = uid.String()
	return nil
```